// accord/lib/adapters/escape-html.js
// Generated by CoffeeScript 1.7.0
(function() {
  var Adapter, EscapeHTML, W, _,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  Adapter = require('../adapter_base');

  W = require('when');

  _ = require('lodash');

  EscapeHTML = (function(_super) {
    var compile;

    __extends(EscapeHTML, _super);

    function EscapeHTML() {
      return EscapeHTML.__super__.constructor.apply(this, arguments);
    }

    EscapeHTML.prototype.name = 'escape-html';

    EscapeHTML.prototype.extensions = ['html'];

    EscapeHTML.prototype.output = 'html';

    EscapeHTML.prototype.supportedEngines = ['he'];

    EscapeHTML.prototype.isolated = true;

    EscapeHTML.prototype._render = function(str, options) {
      options = _.defaults(options, {
        allowUnsafeSymbols: true
      });
      return compile((function(_this) {
        return function() {
          return _this.engine.encode(str, options);
        };
      })(this));
    };

    compile = function(fn) {
      var err, res;
      try {
        res = fn();
      } catch (_error) {
        err = _error;
        return W.reject(err);
      }
      return W.resolve({
        result: res
      });
    };

    return EscapeHTML;

  })(Adapter);

  module.exports = EscapeHTML;

}).call(this);
