// accord/lib/adapters/handlebars.js
// Generated by CoffeeScript 1.7.0
(function() {
  var Adapter, Handlebars, W, fs, path, _,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  Adapter = require('../adapter_base');

  _ = require('lodash');

  path = require('path');

  fs = require('fs');

  W = require('when');

  Handlebars = (function(_super) {
    var compile, register_helpers;

    __extends(Handlebars, _super);

    function Handlebars() {
      return Handlebars.__super__.constructor.apply(this, arguments);
    }

    Handlebars.prototype.name = 'handlebars';

    Handlebars.prototype.extensions = ['hbs', 'handlebars'];

    Handlebars.prototype.output = 'html';

    Handlebars.prototype._render = function(str, options) {
      var compiler;
      compiler = _.clone(this.engine);
      register_helpers(compiler, options);
      return compile((function(_this) {
        return function() {
          return compiler.compile(str)(options);
        };
      })(this));
    };

    Handlebars.prototype._compile = function(str, options) {
      var compiler;
      compiler = _.clone(this.engine);
      register_helpers(compiler, options);
      return compile((function(_this) {
        return function() {
          return compiler.compile(str);
        };
      })(this));
    };

    Handlebars.prototype._compileClient = function(str, options) {
      var compiler;
      compiler = _.clone(this.engine);
      register_helpers(compiler, options);
      return compile((function(_this) {
        return function() {
          return "Handlebars.template(" + (compiler.precompile(str)) + ");";
        };
      })(this));
    };

    Handlebars.prototype.clientHelpers = function() {
      var runtime_path;
      runtime_path = path.join(this.engine.__accord_path, 'dist/handlebars.runtime.min.js');
      return fs.readFileSync(runtime_path, 'utf8');
    };


    /**
     * @private
     */

    register_helpers = function(compiler, opts) {
      if (opts.helpers) {
        compiler.helpers = _.merge(compiler.helpers, opts.helpers);
      }
      if (opts.partials) {
        return compiler.partials = _.merge(compiler.partials, opts.partials);
      }
    };

    compile = function(fn) {
      var err, res;
      try {
        res = fn();
      } catch (_error) {
        err = _error;
        return W.reject(err);
      }
      return W.resolve({
        result: res
      });
    };

    return Handlebars;

  })(Adapter);

  module.exports = Handlebars;

}).call(this);
