// accord/lib/adapters/babel.js
// Generated by CoffeeScript 1.7.0
(function() {
  var Adapter, SixtoFive, W, path, sourcemaps,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  Adapter = require('../adapter_base');

  path = require('path');

  W = require('when');

  sourcemaps = require('../sourcemaps');

  SixtoFive = (function(_super) {
    var compile;

    __extends(SixtoFive, _super);

    function SixtoFive() {
      return SixtoFive.__super__.constructor.apply(this, arguments);
    }

    SixtoFive.prototype.name = 'babel';

    SixtoFive.prototype.extensions = ['jsx'];

    SixtoFive.prototype.output = 'js';

    SixtoFive.prototype.isolated = true;

    SixtoFive.prototype._render = function(str, options) {
      var filename;
      filename = options.filename;
      if (options.sourcemap === true) {
        options.sourceMap = true;
      }
      options.sourceMapName = filename;
      delete options.sourcemap;
      return compile((function(_this) {
        return function() {
          return _this.engine.transform(str, options);
        };
      })(this));
    };

    compile = function(fn) {
      var data, err, res;
      try {
        res = fn();
      } catch (_error) {
        err = _error;
        return W.reject(err);
      }
      data = {
        result: res.code
      };
      if (res.map) {
        return sourcemaps.inline_sources(res.map).then(function(map) {
          data.sourcemap = map;
          return data;
        });
      } else {
        return W.resolve(data);
      }
    };

    return SixtoFive;

  })(Adapter);

  module.exports = SixtoFive;

}).call(this);
