// accord/lib/adapters/csso.js
// Generated by CoffeeScript 1.7.0
(function() {
  var Adapter, CSSO, W,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  Adapter = require('../adapter_base');

  W = require('when');

  CSSO = (function(_super) {
    var compile;

    __extends(CSSO, _super);

    function CSSO() {
      return CSSO.__super__.constructor.apply(this, arguments);
    }

    CSSO.prototype.name = 'csso';

    CSSO.prototype.extensions = ['css'];

    CSSO.prototype.output = 'css';

    CSSO.prototype.isolated = true;

    CSSO.prototype._render = function(str, options) {
      if (options.noRestructure == null) {
        options.noRestructure = false;
      }
      return compile((function(_this) {
        return function() {
          return _this.engine.justDoIt(str, options.noRestructure);
        };
      })(this));
    };

    compile = function(fn) {
      var err, res;
      try {
        res = fn();
      } catch (_error) {
        err = _error;
        return W.reject(err);
      }
      return W.resolve({
        result: res
      });
    };

    return CSSO;

  })(Adapter);

  module.exports = CSSO;

}).call(this);
