// accord/lib/adapters/dogescript.js
// Generated by CoffeeScript 1.7.0
(function() {
  var Adapter, DogeScript, W,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  Adapter = require('../adapter_base');

  W = require('when');

  DogeScript = (function(_super) {
    __extends(DogeScript, _super);

    function DogeScript() {
      return DogeScript.__super__.constructor.apply(this, arguments);
    }

    DogeScript.prototype.name = 'dogescript';

    DogeScript.prototype.extensions = ['djs'];

    DogeScript.prototype.output = 'js';

    DogeScript.prototype.isolated = true;

    DogeScript.prototype._render = function(str, options) {
      return W.resolve({
        result: this.engine(str, options.beauty, options.trueDoge)
      });
    };

    return DogeScript;

  })(Adapter);

  module.exports = DogeScript;

}).call(this);
