// accord/lib/adapters/scss.js
// Generated by CoffeeScript 1.7.0
(function() {
  var Adapter, SCSS, W, path,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  Adapter = require('../adapter_base');

  W = require('when');

  path = require('path');

  SCSS = (function(_super) {
    __extends(SCSS, _super);

    function SCSS() {
      return SCSS.__super__.constructor.apply(this, arguments);
    }

    SCSS.prototype.name = 'scss';

    SCSS.prototype.extensions = ['scss'];

    SCSS.prototype.output = 'css';

    SCSS.prototype.supportedEngines = ['node-sass'];

    SCSS.prototype._render = function(str, options) {
      var deferred;
      deferred = W.defer();
      if (options.sourcemap === true) {
        options.sourceMap = true;
        options.outFile = path.basename(options.filename).replace('.scss', '.css');
        options.omitSourceMapUrl = true;
        options.sourceMapContents = true;
      }
      options.file = options.filename;
      options.data = str;
      options.error = function(res) {
        return deferred.reject({
          result: res
        });
      };
      options.success = function(res) {
        var data;
        data = {
          result: res.css
        };
        if (res.map && Object.keys(JSON.parse(res.map)).length) {
          data.sourcemap = JSON.parse(res.map);
          data.sourcemap.sources.pop();
          data.sourcemap.sources.push(options.file);
        }
        return deferred.resolve(data);
      };
      this.engine.render(options);
      return deferred.promise;
    };

    return SCSS;

  })(Adapter);

  module.exports = SCSS;

}).call(this);
